### PR TITLE
Convert SettingsStore to TypeScript

### PR DESCRIPTION
**Pull request by turt2live**
**This is reviewable commit-by-commit (highly recommended)**

This does not do anything special with the settings, just converts it directly to TypeScript. Fancy `Settings.get(Setting.Breadcrumbs)` types are reserved for another PR.

### CODE REVIEW

File: src/settings/handlers/LocalEchoWrapper.ts [hunk 1]
```diff
@@ -23,47 +23,48 @@ import SettingsHandler from "./SettingsHandler";
...
+        const handlerPromise = this.handler.setValue(settingName, roomId, newValue);
         return Promise.resolve(handlerPromise).finally(() => {
             delete bySetting[cacheRoomId];
         });
     }
 
-
-    canSetValue(settingName, roomId) {
-        return this._handler.canSetValue(settingName, roomId);
+    public canSetValue(settingName, roomId): boolean {
```
Comment (t3chguy) [inline src/settings/handlers/LocalEchoWrapper.ts:63]: string, string

File: src/settings/SettingsStore.ts [hunk 2]
```diff
@@ -61,20 +63,39 @@ for (const key of Object.keys(LEVEL_HANDLERS)) {
...
+    roomId: string,
+    atLevel: SettingLevel,
+    newValAtLevel: any,
+    newVal: any,
+) => void;
+
+interface HandlerMap {
+    // @ts-ignore - TS wants this to be a string key but we know better
+    [level: SettingLevel]: SettingsHandler;
+}
```
Comment (t3chguy) [inline src/settings/SettingsStore.ts:86]: but its called a Map :'(

File: src/settings/SettingsStore.ts [hunk 3]
```diff
@@ -61,20 +63,39 @@ for (const key of Object.keys(LEVEL_HANDLERS)) {
...
 
+export type CallbackFn = (
+    settingName: string,
+    roomId: string,
+    atLevel: SettingLevel,
+    newValAtLevel: any,
+    newVal: any,
+) => void;
+
+interface HandlerMap {
```
Comment (t3chguy) [inline src/settings/SettingsStore.ts:83]: `IHandlerMap`?

File: src/settings/SettingsStore.ts [hunk 4]
```diff
@@ -576,7 +610,7 @@ export default class SettingsStore {
         return handlers;
     }
 
-    static _getFeatureState(settingName) {
+    private static getFeatureState(settingName: string): "labs" | "disable" | "enable" | string {
```
Comment (t3chguy) [inline src/settings/SettingsStore.ts:613]: can we get an Enum for this?

File: src/settings/handlers/SettingsHandler.ts [hunk 5]
```diff
@@ -44,9 +44,9 @@ export default class SettingsHandler {
      * @param {*} newValue The new value for the setting, may be null.
      * @returns {Promise} Resolves when the setting has been saved.
      */
-    setValue(settingName, roomId, newValue) {
+    public setValue(settingName: string, roomId: string, newValue: any): Promise<void> {
```
Comment (t3chguy) [inline src/settings/handlers/SettingsHandler.ts:47]: abstract?

File: src/settings/SettingsStore.ts (hunk 2 above)
Comment (turt2live) [reply src/settings/SettingsStore.ts:86]: refactoring all of settings to use maps is a different PR's problem.

File: src/settings/SettingsStore.ts (hunk 4 above)
Comment (turt2live) [reply src/settings/SettingsStore.ts:613]: I'll declare it as a `type`. It's not going to survive long.

File: src/settings/WatchManager.ts [hunk 6]
```diff
@@ -14,41 +14,52 @@ See the License for the specific language governing permissions and
...
+}
+
 /**
  * Generalized management class for dealing with watchers on a per-handler (per-level)
  * basis without duplicating code. Handlers are expected to push updates through this
  * class, which are then proxied outwards to any applicable watchers.
  */
 export class WatchManager {
-    _watchers = {}; // { settingName: { roomId: callbackFns[] } }
+    private watchers: {[settingName: string]: RoomWatcherMap} = {};
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/settings/WatchManager.ts:34]: not refactoring settings to use maps here.

File: src/settings/handlers/LocalEchoWrapper.ts [hunk 7]
```diff
@@ -23,47 +23,48 @@ import SettingsHandler from "./SettingsHandler";
  * handler as much as possible to ensure values are not stale.
  */
 export default class LocalEchoWrapper extends SettingsHandler {
+    private cache: {
+        [settingName: string]: {
+            [roomId: string]: any;
+        };
+    } = {};
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/settings/handlers/LocalEchoWrapper.ts:30]: still not refactoring settings to use them here

**Pull request merged**